File: CardMasterSkin/Elements/SEImage.vb
Imports System.Drawing
Imports System.IO
Imports System.Reflection
Imports CardMasterCard.Card
Imports CardMasterSkin.GraphicsElement


Namespace Skins

    Public Enum ResourceTypes
        Image
        Texture
    End Enum

    Public Class SEImage
        Inherits SkinElement

        Public Property ImageName As String = Nothing
        Public Property NameAttribute As String
        Public Property ResourceType As ResourceTypes = ResourceTypes.Texture

        Public Sub New(skin As Skin, width As Integer, height As Integer)
            Me.New(skin, 0, 0, width, height, Nothing)
        End Sub

        Public Sub New(skin As Skin, width As Integer, height As Integer, imageName As String)
            Me.New(skin, 0, 0, width, height, imageName)
        End Sub

        Public Sub New(skin As Skin, x As Integer, y As Integer, width As Integer, height As Integer)
            Me.New(skin, x, y, width, height, Nothing)
        End Sub

        Public Sub New(skin As Skin, x As Integer, y As Integer, width As Integer, height As Integer, imageName As String)
            MyBase.New(skin, x, y, width, height)

            Me.ImageName = imageName

        End Sub

        Protected Overrides Function GetGraphicElements(card As Card) As List(Of GraphicElement)

            Dim graphicElementsList As New List(Of GraphicElement)()

            Dim fileFullname = GetFileFullname(card)

            If fileFullname IsNot Nothing Then

                Dim img As Image = Bitmap.FromFile(fileFullname, True)

                graphicElementsList.Add(New ImageElement(img, New Rectangle(Me.X, Me.Y, Me.Width, Me.Height)))

            End If

            Return graphicElementsList

        End Function

        Private Function GetFileFullname(card As Card) As String

            Dim fileFullname As String = Nothing

            Dim rootDirectory As DirectoryInfo = Nothing

            Select Case Me.ResourceType
                Case ResourceTypes.Image : rootDirectory = m_skin.ImagesDirectory
                Case Else : rootDirectory = m_skin.TexturesDirectory
            End Select


            If (RootDirectory IsNot Nothing) AndAlso (RootDirectory.Exists) Then

                Dim fileName As String = Nothing

                If Not String.IsNullOrEmpty(Me.NameAttribute) Then

                    Dim cardType As Type = GetType(Card)
                    Dim cardProperty As PropertyInfo = cardType.GetProperty(Me.NameAttribute)
                    Dim cardGetMethod As MethodInfo = cardProperty.GetGetMethod
                    Dim propertyValue As Object = cardGetMethod.Invoke(card, Nothing)

                    If propertyValue IsNot Nothing Then
                        fileName = propertyValue.ToString
                    End If

                Else
                    fileName = ImageName
                End If

                If fileName IsNot Nothing Then

                    fileName = fileName.Replace(":", "-").Replace("""", "'")

                    With rootDirectory.GetFiles(fileName & ".*", SearchOption.AllDirectories)
                        If .Count > 0 Then
                            fileFullname = .First.FullName
                        End If
                    End With
                End If

            End If

            Return fileFullname

        End Function

    End Class

End Namespace

File: CardMasterSkin/Elements/SEImage.cs
using CardMasterCard.Card;
using CardMasterSkin.GraphicsElements;
using CardMasterSkin.Skins;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Reflection;

namespace CardMasterSkin.Elements
{
    public class SEImage : SkinElement
    {
        public string ImageName { get; set; } = null;
        public string NameAttribute { get; set; } = null;

        public SEImage(Skin skin, int x, int y, int width, int height) : base(skin, x, y, width, height)
        { }

        public SEImage(Skin skin, int x, int y, int width, int height, string imageName) : base(skin, x, y, width, height)
        {
            this.ImageName = imageName;
        }

        protected override List<GraphicElement> GetGraphicElements(Card card)
        {
            var graphicElementsList = new List<GraphicElement>();
            string fileFullname = GetFileFullname(card);

            if (fileFullname != null)
            {
                Image img = Bitmap.FromFile(fileFullname, true);
                graphicElementsList.Add(new ImageElement(img, new Rectangle(this.X, this.Y, this.Width, this.Height)));
            }

            return graphicElementsList;

        }

        private string GetFileFullname(Card card)
        {
            string fileFullname = null;
            DirectoryInfo rootDirectory = this.skin.ResourcesDirectory;

            if ((rootDirectory != null) && (rootDirectory.Exists))
            {
                string fileName = null;

                if (!string.IsNullOrEmpty(this.NameAttribute))
                {
                    Type cardType = card.GetType();
                    PropertyInfo cardProperty = cardType.GetProperty(this.NameAttribute);
                    MethodInfo cardGetMethod = cardProperty.GetGetMethod();
                    object propertyValue = cardGetMethod.Invoke(card, null);

                    if (propertyValue != null)
                    {
                        fileName = propertyValue.ToString();
                    }

                }
                else
                {
                    fileName = ImageName;
                }

                if (fileName != null)
                {
                    fileName = fileName.Replace(":", "-").Replace("\"", "'");
                    FileInfo[] files = rootDirectory.GetFiles(fileName + ".*", SearchOption.AllDirectories);

                    if (files.Length > 0)
                    {
                        fileFullname = files[0].FullName;
                    }

                    files = null;
                }

            }

            return fileFullname;

        }

    }
}
